src/printer_tree.c

```diff
@@ -0,0 +1,2757 @@
+        return TRD_FLAGS_TYPE_RO;
+    } else if (nodetype & LYS_NOTIF) {
+        return TRD_FLAGS_TYPE_NOTIF;
+    } else if (nodetype & LYS_USES) {
+        return TRD_FLAGS_TYPE_USES_OF_GROUPING;
+    } else if (nodetype & (LYS_RPC | LYS_ACTION)) {
+        return TRD_FLAGS_TYPE_RPC;
+        /* if config is not set then look at ancestor's config and get his config */
+    } else if (!(flags & (LYS_CONFIG_R | LYS_CONFIG_W))) {
+        return tro_lysp_flags2config(ca_lys_config);
+    } else {
+        return tro_lysp_flags2config(flags);
+    }
+}
+
+static trt_node_type
+tro_resolve_node_type(const struct lysp_node *pn, const struct lysp_node_list *ca_last_list)
+{
+    if (pn->nodetype & (LYS_INPUT | LYS_OUTPUT)) {
+        return TRD_NODE_ELSE;
+    } else if (pn->nodetype & LYS_CASE) {
+        return TRD_NODE_CASE;
+    } else if (pn->nodetype & LYS_CHOICE && !(pn->flags & LYS_MAND_TRUE)) {
+        return TRD_NODE_OPTIONAL_CHOICE;
+    } else if (pn->nodetype & LYS_CHOICE) {
+        return TRD_NODE_CHOICE;
+    } else if (pn->nodetype & LYS_CONTAINER && tro_lysp_container_has_presence(pn)) {
+        return TRD_NODE_CONTAINER;
+    } else if (pn->nodetype & LYS_LIST && tro_lysp_list_has_keys((const struct lysp_node_list *)pn)) {
+        return TRD_NODE_KEYS;
+    } else if (pn->nodetype & (LYS_LIST | LYS_LEAFLIST)) {
+        return TRD_NODE_LISTLEAFLIST;
+    } else if (pn->nodetype & (LYS_ANYDATA | LYS_ANYXML) && !(pn->flags & LYS_MAND_TRUE)) {
+        return TRD_NODE_OPTIONAL;
+    } else if (pn->nodetype & LYS_LEAF && !(pn->flags & LYS_MAND_TRUE) && !tro_lysp_leaf_is_key(pn, ca_last_list)) {
+        return TRD_NODE_OPTIONAL;
+    } else {
+        return TRD_NODE_ELSE;
+    }
+}
+
+/** Transformation of current lysp_node to struct trt_node. */
+static struct trt_node
+tro_read_node(struct trt_parent_cache ca, const struct trt_tree_ctx *tc)
+{
+    assert(tc != NULL && tc->pn != NULL);
+    const struct lysp_node *pn = tc->pn;
+
+    assert(pn->nodetype != LYS_UNKNOWN);
+
+    struct trt_node ret = trp_empty_node();
+
+    /* remember:
+     * - LYS_INTPUT and LYS_OUTPUT lysp_node don't have
+     *      flags, name and iffeatures element in their structure.
+     */
+
+    /* <status> */
+    ret.status = tro_resolve_status(pn->nodetype, pn->flags, ca.lys_status);
+
+    /* TODO: TRD_FLAGS_TYPE_MOUNT_POINT aka "mp" is not supported right now. */
+    /* <flags> */
+    ret.flags = tro_resolve_flags(pn->nodetype, pn->flags, ca.ancestor, ca.lys_config);
+
+    /* TODO: TRD_NODE_TOP_LEVEL1 aka '/' is not supported right now. */
+    /* TODO: TRD_NODE_TOP_LEVEL2 aka '@' is not supported right now. */
+    /* set type of the node */
+    ret.name.type = tro_resolve_node_type(pn, ca.last_list);
+
+    /* TODO: ret.name.module_prefix is not supported right now. */
+    ret.name.module_prefix = NULL;
+
+    /* set node's name */
+    ret.name.str =
+            pn->nodetype & (LYS_INPUT) ? "input" :
```

Comment:
it is not necessary to handle this now, the nodes have the "input"/"output" name now (if not, it is a bug)